rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+#### Flexibility for collaborators
+
+Currently, any partner or contributor that would like to work with us are subject to all the rules within of the main repository. Some of these can be relaxed through modularization, where work can happen in a separate repository.
+
+#### Large TF support matrix
+
+The TF support story is a unique beast. Our support matrix has a lot of orthogonal dimensions. Below are some of the more prominent dimensions:
+
+*   Environment (google3, opensource)
+*   Operating system (Linux, windows, macos, mobile)
+*   Architecture (x86, arm, ppc64, …)
+*   Accelerator (CPU, GPU,  TPU)
+*   Compiler (GCC, Clang, MSVC)
+*   Python version (2, 3.4, 3.5, 3.6,  3.7)
+
+More can be added to this list where we have to rebuild TensorFlow to support different network architectures, CUDA versions, SIMD instruction sets, etc.
+
+Having a monolithic repository means we need to rebuild all of our code for all of these different combinations. However, it makes no sense to rebuild all of our C++ code if the only difference is the Python version. Or rebuild all of our CPU kernels for different CUDA versions. Modularizing our code means we only need to rebuild and test the modules that are directly impacted by the dimensions we are changing in the support matrix.
+
+
+## Overview
+
+This project aims to split the TensorFlow codebase into **smaller, more focused**, repositories that can be released and managed separately. These modules will talk to each other using **well defined APIs** that guarantee backwards compatibility. Thanks to these APIs, these modules will be **managed/owned/released independently**. There will be different strategies to break apart pieces based on the languages, but below summarizes the approach for C++ and Python:
+
+
+![alt_text](20190305-modular-tensorflow/big_picture.png "Overview of modular TensorFlow")
+
+A summary of the above is:
+
+
+
+*   Core TF functionality will be implemented in C++
+*   Core TF functionality can be extended using shared objects.
+*   On top of the core C++ libraries, we will have the language bindings (Using the C API)
+*   There can be more functionality built on top of the core TF bindings in different languages, which can be maintained and distributed separately.
+*   All different pieces need to use Stable public APIs with backwards compatibility guarantees.
+
+
+### Definitions
+
+This section will briefly describe the terms we will use in the rest of this design.
+
+**Modules:** These are the components of core TF library that will "accept" plugins to expand their capabilities. Examples of modules are networking, filesystem, (graph) optimizers.
+
+**Plugins:** Plugins are extensions to different modules. For example, filesystem module can have a GCP plugin, an S3 plugin, an HDFS plugin.
```

Comment:
This is defining an SDK. When can we get drop? That will generate lots of feedback/questions...

Reply:
@sjamesr is working on this. We are trying to build the API based on `OpKernel` and `OpKernelContext` classes.

Reply:
A further clarification.
The SDK will start as what it is today.
We will try to make all the kernel code work as is. Therefore, you can think of the current TF C++ headers as v0 of what is to come for TF C++ APIs.